### ISSUE TITLE
is that possible to replace your TrilinearInterpolation with torch.nn.functional.grid_sample?

### ISSUE THREAD

**Issue by Dongshengjiang** [HuiZeng/Image-Adaptive-3DLUT#14]
is that possible to replace your TrilinearInterpolation with torch.nn.functional.grid_sample? If ok, it would be more convenient.

**Comment by HuiZeng** [comment]
Hi, this is an interesting idea. 
But it seems not that easy since their interpolation details are different.

**Comment by tuxa** [comment]
Hi! grid_sample actually uses trilinear interpolation if a 5d tensor is passed when mode is set to bilinear. 

So, instead of using 
```
_, result = trilinear_(LUT, img)
```
one can use:

```
# scale im between -1 and 1 since its used as grid input in grid_sample
img = (img - .5) * 2.

# grid_sample expects NxDxHxWx3 (1x1xHxWx3)
img = img.permute(0, 2, 3, 1)[:, None]

# add batch dim to LUT
LUT = LUT[None]

# grid sample
result = F.grid_sample(LUT, img, mode='bilinear', padding_mode='border', align_corners=True)

# drop added dimensions and permute back to 1x3xHxW
result = result[:, :, 0].permute(0, 2, 3, 1)

```

**Comment by coallar** [comment]
changing to:ndarr = result.squeeze().mul_(255).add_(0.5).clamp_(0, 255).permute(1, 2, 0).to('cpu', torch.uint8).numpy() to ndarr = result.squeeze().mul_(255).add_(0.5).clamp_(0, 255).to('cpu', torch.uint8).numpy()
is necessary

**Comment by ZedFu** [comment]
Has anyone ever used grid_sample instead of Trilinear Interpolation? Why is the result of using grid_sample worse?

**Comment by 11923303233** [comment]
# scale im between -1 and 1 since its used as grid input in grid_sample
img = (img - .5) * 2.

# grid_sample expects NxDxHxWx3 (1x1xHxWx3)
img = img.permute(0, 2, 3, 1)[:, None]

# add batch dim to LUT
LUT = LUT[None]

# grid sample
result = F.grid_sample(LUT, img, mode='bilinear', padding_mode='border', align_corners=True)

# drop added dimensions and permute back
result = result[:, :, 0].permute(0, 2, 3, 1)



How to set batch_size > 1?    LUT = LUT[None]  --> LUT = LUT[None].repeat(bs, 1,1,1,1) is ok?

**Comment by fedral** [comment]
> Hi! grid_sample actually uses trilinear interpolation if a 5d tensor is passed when mode is set to bilinear.
> 
> So, instead of using
> 
> ```
> _, result = trilinear_(LUT, img)
> ```
> 
> one can use:
> 
> ```
> # scale im between -1 and 1 since its used as grid input in grid_sample
> img = (img - .5) * 2.
> 
> # grid_sample expects NxDxHxWx3 (1x1xHxWx3)
> img = img.permute(0, 2, 3, 1)[:, None]
> 
> # add batch dim to LUT
> LUT = LUT[None]
> 
> # grid sample
> result = F.grid_sample(LUT, img, mode='bilinear', padding_mode='border', align_corners=True)
> 
> # drop added dimensions and permute back
> result = result[:, :, 0].permute(0, 2, 3, 1)
> ```

good trick!

**Comment by gganyy** [comment]
> Hi! grid_sample actually uses trilinear interpolation if a 5d tensor is passed when mode is set to bilinear.
> 
> So, instead of using
> 
> ```
> _, result = trilinear_(LUT, img)
> ```
> 
> one can use:
> 
> ```
> # scale im between -1 and 1 since its used as grid input in grid_sample
> img = (img - .5) * 2.
> 
> # grid_sample expects NxDxHxWx3 (1x1xHxWx3)
> img = img.permute(0, 2, 3, 1)[:, None]
> 
> # add batch dim to LUT
> LUT = LUT[None]
> 
> # grid sample
> result = F.grid_sample(LUT, img, mode='bilinear', padding_mode='border', align_corners=True)
> 
> # drop added dimensions and permute back
> result = result[:, :, 0].permute(0, 2, 3, 1)
> ```

/
/

> Hi, this is an interesting idea. But it seems not that easy since their interpolation details are different.

Could you please add an _ext function to the code? it seemed that the code lack it.

**Comment by zzd1992** [comment]
Does any one test and compare the speed of both methods?
